### ISSUE TITLE
CanMatch and UrlMatcher is called twice

### ISSUE THREAD

**Issue by dawidgarus** [angular/angular#26081]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
`UrlMatcher` function is called twice on navigation

## Expected behavior
`UrlMatcher` should be called once

## Minimal reproduction of the problem with instructions
https://stackblitz.com/edit/angular-router-basic-example-wdvkbt?file=app/app.routing.module.ts
As You can see, on each navigation `UrlMatcher` is called twice. What's even weirder is that `group` argument is different each time - one has `parent` and one doesn't.

## What is the motivation / use case for changing the behavior?
Originally I wanted to create matcher that would match route only if You are in specific named outlet:
```typescript
export function parentOutletMatcher(outlet: string): UrlMatcher {
	return (segments: UrlSegment[], group: UrlSegmentGroup, route: Route) => {
		if (group.parent && group.parent.children[outlet] === group) {
			return { consumed: [] };
		}
		return null;
	};
}
```
So I could use one lazy-loaded module to define routes for many named outlets, example:
```typescript
const ROOT_ROUTES = [
  { path: 'feature', loadChildren: './lazy.module.ts#LazyModule' },
  { path: 'feature', outlet: 'modal', loadChildren: './lazy.module.ts#LazyModule' },
  { path: 'feature', outlet: 'settings', loadChildren: './lazy.module.ts#LazyModule' },
];
// in lazy.module.ts:
const LAZY_ROUTES = [
  { matcher: parentOutletMatcher(PRIMARY_OUTLET), children: [
    // ...
  ]},
  { matcher: parentOutletMatcher('modal'), children: [
    // ...
  ]},
  { matcher: parentOutletMatcher('settings'), children: [
    // ...
  ]},
]
```
But because of router's weird behavior this doesn't work.

## Environment

<pre><code>
Angular version: 6.1.8

Browser:
- [x] Chrome (desktop) version XX
</code></pre>

**Comment by sambaptista** [comment]
Still happening in angular 8.2

**Comment by Eseperio** [comment]
Also happens in 9.1.4

**Comment by uragun91** [comment]
Any update on this?

**Comment by basherr** [comment]
Still getting the same issue on **Angular 10.0.3** on Windows 10

**Comment by Parsa-Sedigh** [comment]
Have the same problem in angular 11.2.8

**Comment by huynv86** [comment]
same problem on version 12.1.0

**Comment by antonjstu** [comment]
Still happens on Angular 10.2.7, any updates on this?

**Comment by theideasaler** [comment]
Still an issue in Angular14

**Comment by farshadhemmati** [comment]
It is calling it 3 times in Angular 14

**Comment by robertandreimoldoveanu** [comment]
I encountered 4 calls in Angular 15. FWIW, two seem to come from the `apply` call in `packages/router/src/apply_redirects.ts` , then there's a call to the `canMatch` function, and then two more calls from the `recognise` call in `packages/router/src/recognize.ts`.

**Comment by HyperLife1119** [comment]
The CanMatch route guard is also called twice, is that related to this issue?

**Comment by csvn** [comment]
I think I get the same issue described regarding `CanMatch`. It seems to happen during/after the `NavigationStart` and `RouteConfigLoadEnd` events.

![image](https://user-images.githubusercontent.com/8770194/217609713-2ca321b9-602f-408a-a061-abb117c4618a.png)

In our project we need to make an asynchronous request (checking version of a service) before deciding to show two different views. This cannot be done as cleanly via other methods than `CanMatch`; showing view A or B. Unfortunately, this issue means I need need to cache the result to avoid sending multiple identical requests.

Would be a very nice quality of life to ensure this method is only called once, if possible.

**Comment by atscott** [comment]
The goal is to have this addressed in V16. This is a difficult change to make because it requires completely rewriting parts of the URL creation and matching logic in the Router internals. The changes required have spanned multiple versions so far and include multiple breaking changes. The approach here has been cautious and slow to avoid breaking applications in hard to debug ways. 

* The first change required was released in 14.1: https://github.com/angular/angular/pull/45877 is the new logic for generating URLs based on an `ActivatedRouteSnapshot` rather than the current URL tree
* The first breaking change was made in v15 https://github.com/angular/angular/pull/47623. The new approach does not support an old bug that had an opt-in.
* V16 will switch the default behavior for generating URLs in https://github.com/angular/angular/pull/48508. This is also a breaking change because it can result in new (but correct) routing behavior in applications.
* After the above is merged, we can finally start combining the first two stages in the Router into one. These stages duplicate the route matching logic, which requires running the matchers twice.
